runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1175,3 +1176,163 @@ J9::ValuePropagation::getParmValues()
+         return true;
+      }
+
+   return false;
+   }
+
+
+// Do not add fear point in a frame that doesn't support OSR
+static bool safeToAddFearPointIn(TR::Compilation* comp, int32_t callerIndex)
+   {
+   TR::ResolvedMethodSymbol *method = callerIndex == -1 ?
+      comp->getJittedMethodSymbol() : comp->getInlinedResolvedMethodSymbol(callerIndex);
+
+   return !method->cannotAttemptOSRDuring(callerIndex, comp, false);
+   }
+
+bool J9::ValuePropagation::transformDirectLoad(TR::Node* node)
+   {
+   // Allow OMR to fold reliable static final field first
+   if (OMR::ValuePropagation::transformDirectLoad(node))
+      return true;
+
+   if (node->isLoadOfStaticFinalField() &&
+       tryFoldStaticFinalFieldAt(_curTree, node))
+      {
+      return true;
+      }
+
+   return false;
+   }
+
+/** \brief
+ *     Try to fold static final field
+ *
+ *  \param tree
+ *     The tree with the load of static final field.
+ *
+ *  \param fieldNode
+ *     The node which is a load of a static final field.
+ */
+bool J9::ValuePropagation::tryFoldStaticFinalFieldAt(TR::TreeTop* tree, TR::Node* fieldNode)
+   {
+   TR_ASSERT(fieldNode->isLoadOfStaticFinalField(), "Node n%dn %p has to be a load of a static final field", fieldNode->getGlobalIndex(), fieldNode);
+
+   if (comp()->getOption(TR_DisableGuardedStaticFinalFieldFolding))
+      {
+      return false;
+      }
+
+   if (!comp()->supportsInduceOSR() ||
+       !comp()->isOSRTransitionTarget(TR::postExecutionOSR) ||
+       !comp()->getOSRMode() == TR::voluntaryOSR)
+      {
+      return false;
+      }
+
+   if (fieldNode->getByteCodeInfo().doNotProfile() ||
+       skipFinalFieldFoldingInBlock(tree->getEnclosingBlock()) ||
+       !safeToAddFearPointIn(comp(), fieldNode->getByteCodeInfo().getCallerIndex()) ||
+       TR::TransformUtil::canFoldStaticFinalField(comp(), fieldNode) != TR_maybe)
+      {
+      return false;
+      }
+
+   TR::SymbolReference* symRef = fieldNode->getSymbolReference();
+   if (symRef->hasKnownObjectIndex())
+      {
+      return false;
+      }
+
+   int32_t cpIndex = symRef->getCPIndex();
+   TR_OpaqueClassBlock* declaringClass = symRef->getOwningMethod(comp())->getClassFromFieldOrStatic(comp(), cpIndex);
+   int32_t fieldNameLen;
+   char* fieldName = symRef->getOwningMethod(comp())->fieldName(cpIndex, fieldNameLen, comp()->trMemory(), stackAlloc);
+   int32_t fieldSigLength;
+   char* fieldSignature = symRef->getOwningMethod(comp())->staticSignatureChars(cpIndex, fieldSigLength);
+
+   if (trace())
+      {
+      traceMsg(comp(),
+              "Looking at static final field n%dn %.*s declared in class %p\n",
+              fieldNode->getGlobalIndex(), fieldNameLen, fieldName, declaringClass);
+      }
+
+   if (isStaticFinalFieldWorthFolding(comp(), declaringClass, fieldSignature, fieldSigLength))
+      {
+      if (TR::TransformUtil::foldStaticFinalFieldAssumingProtection(comp(), fieldNode))
```

Comment:
A perform transformation would seem to be a good idea here...

Reply:
`foldStaticFinalFieldAssumingProtection` does a perform transformation.